Framework/API/src/ExperimentInfo.cpp

```diff
@@ -892,7 +881,41 @@ ExperimentInfo::getInstrumentFilename(const std::string &instrumentName,
   return mostRecentIDF;
 }
 
-//--------------------------------------------------------------------------------------------
+/** Return a const reference to the DetectorInfo object.
+ *
+ * Any modifications of the instrument or instrument parameters will invalidate
+ * this reference.
+ */
+const DetectorInfo &ExperimentInfo::detectorInfo() const {
+  if (!m_detectorInfo) {
+    std::lock_guard<std::mutex> lock{m_detectorInfoMutex};
```

Comment:
The method is `const` so I wanted to make it thread-safe. Admittedly, `MatrixWorkspace::detectorInfo()` should typically be called outside threaded contexts, but there are exceptions.